Do not apply user-coupled promotions to all carts

mvz opened the pull request:
The code in the Cart promotion handler will first try if a promotion is eligible for the line item. However, Promotion#eligible? is implemented in such a way that it will consider itself eligible for a line item
if it is only applicable to orders.

Therefore, a user-related promotion without a promotion code, which is not applicable to line items, will be considered eligible for all orders. If there are many such promotions, this means each order will be burdened by a large number of inactive promotion adjustments.

This change fixes this problem by checking for eligibility with both line item and order.

Closes #7227.

mvz committed:
Do not apply user-coupled promotions to all carts

The code in the Cart promotion handler will first try if a promotion is
eligible for the line item. However, Promotion#eligible? is implemented
in such a way that it will consider itself eligible for a line item
if it is only applicable to orders.

Therefore, a user-related promotion without a promotion code, which
is not applicable to line items, will be considered eligible for all
orders. If there are many such promotions, this means each order will be
burdened by a large number of inactive promotion adjustments.

This change fixes this problem by checking for eligibility with both
line item and order.

mvz commented:
Remaining failing tests seem to be unrelated.

MatthewKennedy commented:
I think I ran into this recently, the cart would lock up with this error message"
![cart error](https://user-images.githubusercontent.com/1240766/51804923-a3affa00-225e-11e9-96b9-55150a52dfc8.jpg)
I had trade discount codes that were implemented on a user basis plus cart value, going from around 20% up to 40%, but also other codes that none trade users can access that must be adding up altogether in the background.

mvz commented:
I rebased and now all builds are green :tada:.

chhga commented on core/app/models/spree/promotion.rb:
```diff
@@ -120,7 +120,12 @@ def deactivate(payload)
+        !!eligible_rules(promotable, {})
+      end
     end
 
```
let's use early return here. Using `case` when we have only one option looks redundant.

chhga replied:
```suggestion
def eligible?(promotable)
    return false if expired? || usage_limit_exceeded?(promotable) || blacklisted?(promotable)
    return !!eligible_rules(promotable, {}) && !!eligible_rules(promotable.order, {}) if promotable == Spree::LineItem
    
    !!eligible_rules(promotable, {})
end
```

mvz replied:
That makes sense. I'll rewrite this.

chhga commented on core/app/models/spree/promotion.rb:
```diff
@@ -120,7 +120,12 @@ def deactivate(payload)
     def eligible?(promotable)
```
```suggestion
def eligible?(promotable)
    return false if expired? || usage_limit_exceeded?(promotable) || blacklisted?(promotable)
    return !!eligible_rules(promotable, {}) && !!eligible_rules(promotable.order, {}) if promotable == Spree::LineItem
    
    !!eligible_rules(promotable, {})
end
```

mvz replied:
`promotable` is not a class, so comparing it to `Spree::LineItem` using `==` won't work.

mvz commented:
@chhga Somehow I totally missed your review. Taking a look now.

mvz committed:
Replace case statement with simple conditional

mvz commented:
Updated and rebased

mvz committed:
Allow rspec matchers to show both expected and actual values

By moving the indication of promo sequence to a context description, the
default matcher messages can be used, so failure messages will show the
actual values found, improving their helpfulness.

mvz committed:
Adjust expectation because ineligible adjustments are no longer created

mvz commented:
@chhga can you do another review please?

mvz commented:
Unfortunately I no longer have the time nor the need to bring this pull request into a state where it can be merged.
